In [22]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings, os, sys, shutil
warnings.filterwarnings("ignore")

In [23]:
train = pd.read_csv('./icr-identify-age-related-conditions/train.csv')
test = pd.read_csv('./icr-identify-age-related-conditions/test.csv')
meta = pd.read_csv('./icr-identify-age-related-conditions/greeks.csv')

In [24]:
train_init = train.copy()
train['Alpha'] = meta['Alpha']
train['Beta'] = meta['Beta']
train['Gamma'] = meta['Gamma']
train['Delta'] = meta['Delta']

In [25]:
# 处理greeks
# Alpha为A的时候为0，其余为1
train['Alpha'] = train['Alpha'].apply(lambda x: 0 if x == 'A' else 1)

In [26]:
# 将数据集中唯一的离散特征转为01，这可能表明患者的性别
train['EJ'] = train['EJ'].replace({'A': 0, 'B': 1}).astype(float)
test['EJ']  = test['EJ'].replace({'A': 0, 'B': 1}).astype(float)
train_init['EJ'] = train_init['EJ'].replace({'A': 0, 'B': 1}).astype(float)
# 缺失值处理，用中位数填充
train['BQ'].fillna(0, inplace=True)
train.fillna(train.median(), inplace=True)
test.fillna(test.median(), inplace=True)
train_init.fillna(train_init.median(), inplace=True)
# 移除ID列
train_id = train['Id'].copy()
test_id  =  test['Id'].copy()
train = train.drop(['Id'], axis=1)
test  =  test.drop(['Id'], axis=1)

In [27]:
# 将gamma中的M和N都转为0，G，H，E，F，A，B的转为1
train['Gamma'] = train['Gamma'].replace({'M': 0, 'N': 0, 'G': 1, 'H': 1, 'E': 1, 'F': 1, 'A': 1, 'B': 1})
# 将Beta中的C0,B1,A2
train['Beta'] = train['Beta'].replace({'C': 0, 'B': 1, 'A': 2})
# 将Delta中的B0,A1,C1,D2
train['Delta'] = train['Delta'].replace({'B': 0, 'A': 1, 'C': 1, 'D': 2})

train.head()


AB          AF          AH         AM        AR        AX        AY  \
0  0.209377  3109.03329   85.200147  22.394407  8.138688  0.699861  0.025578   
1  0.145282   978.76416   85.200147  36.968889  8.138688  3.632190  0.025578   
2  0.470030  2635.10654   85.200147  32.360553  8.138688  6.732840  0.025578   
3  0.252107  3819.65177  120.201618  77.112203  8.138688  3.685344  0.025578   
4  0.380297  3733.04844   85.200147  14.103738  8.138688  3.942255  0.054810   

          AZ          BC         BD   ...          GE            GF  \
0   9.812214    5.555634  4126.58731  ...   72.611063   2003.810319   
1  13.517790    1.229900  5496.92824  ...   72.611063  27981.562750   
2  12.824570    1.229900  5135.78024  ...   88.609437  13676.957810   
3  11.053708    1.229900  4169.67738  ...   82.416803   2094.262452   
4   3.396778  102.151980  5728.73412  ...  146.109943   8524.370502   

          GH         GI         GL  Class  Alpha  Beta  Gamma  Delta  
0  22.136229  69.834944   0.120343      1      1     0      1      2  
1  29.135430  32.131996  21.978000      0      0     0      0      0  
2  28.022851  35.192676   0.196941      0      0     0      0      0  
3  39.948656  90.493248   0.155829      0      0     0      0      0  
4  45.381316  36.262628   0.096614      1      1     1      1      0  

[5 rows x 61 columns]

In [28]:
# 连续变量归一化处理
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
numeric_columns = [_ for _ in train.columns if _ not in ['EJ', 'Alpha', 'Beta', 'Gamma', 'Delta', 'Class']]
train[numeric_columns] = scaler.fit_transform(train[numeric_columns])
test[numeric_columns] = scaler.transform(test[numeric_columns])
train.head()

AB        AF        AH        AM        AR        AX        AY  \
0 -0.572153 -0.170975 -0.261669 -0.237889 -0.189295 -1.900558 -0.083417   
1 -0.709105 -1.097801 -0.261669 -0.028701 -0.189295 -0.750457 -0.083417   
2 -0.015212 -0.377169 -0.261669 -0.094845 -0.189295  0.465662 -0.083417   
3 -0.480851  0.138196  0.012347  0.547477 -0.189295 -0.729610 -0.083417   
4 -0.206946  0.100517 -0.261669 -0.356885 -0.189295 -0.628845 -0.013229   

         AZ        BC       BD   ...        GE        GF        GH        GI  \
0 -0.173502 -0.038354 -0.405383  ... -0.410260 -0.655511 -0.948991  0.531241   
1  0.678919 -0.104787  0.048541  ... -0.410260  0.687893 -0.238862 -0.509218   
2  0.519453 -0.104787 -0.071089  ... -0.299210 -0.051850 -0.351743 -0.424754   
3  0.112088 -0.104787 -0.391109  ... -0.342195 -0.650833  0.858232  1.101332   
4 -1.649292  1.445139  0.125327  ...  0.099920 -0.318309  1.409422 -0.395228   

         GL  Class  Alpha  Beta  Gamma  Delta  
0 -0.814049      1      1     0      1      2  
1  1.304860      0      0     0      0      0  
2 -0.806623      0      0     0      0      0  
3 -0.810609      0      0     0      0      0  
4 -0.816349      1      1     1      1      0  

[5 rows x 61 columns]

In [29]:
# 抛弃高度相关的特征
drop_cols = ['BZ','CL','EH','GL']  # 高度相关
drop_cols2 = ['DY','CB','GB','CH','DL','CU','FS','AZ','GE','EG','EP']  # 无用特征，同一点取得极值
drop_cols3 = ['BZ', 'DV', 'EH', 'FD ']
train.drop(drop_cols, axis=1, inplace=True)
test.drop(drop_cols, axis=1, inplace=True)

In [30]:
# ytrain为离散
ytrain = train[['Class', 'Alpha', 'Beta', 'Gamma', 'Delta']]
train.drop(['Class', 'Alpha', 'Beta', 'Gamma', 'Delta'], axis=1, inplace=True)

In [31]:
train.head()

AB        AF        AH        AM        AR        AX        AY  \
0 -0.572153 -0.170975 -0.261669 -0.237889 -0.189295 -1.900558 -0.083417   
1 -0.709105 -1.097801 -0.261669 -0.028701 -0.189295 -0.750457 -0.083417   
2 -0.015212 -0.377169 -0.261669 -0.094845 -0.189295  0.465662 -0.083417   
3 -0.480851  0.138196  0.012347  0.547477 -0.189295 -0.729610 -0.083417   
4 -0.206946  0.100517 -0.261669 -0.356885 -0.189295 -0.628845 -0.013229   

         AZ        BC       BD   ...        FE        FI        FL        FR  \
0 -0.173502 -0.038354 -0.405383  ... -0.112922 -2.226608  0.162821 -0.035806   
1  0.678919 -0.104787  0.048541  ... -0.311056  0.084542 -0.457925 -0.060566   
2  0.519453 -0.104787 -0.071089  ... -0.173811  0.517060  0.198663 -0.051023   
3  0.112088 -0.104787 -0.391109  ...  0.058201  1.617160  0.060364 -0.060566   
4 -1.649292  1.445139  0.125327  ...  0.520331  1.212850  0.237302  0.896815   

         FS        GB        GE        GF        GH        GI  
0 -0.250437 -0.940094 -0.410260 -0.655511 -0.948991  0.531241  
1  0.113640 -1.145070 -0.410260  0.687893 -0.238862 -0.509218  
2  0.597343  1.637944 -0.299210 -0.051850 -0.351743 -0.424754  
3 -0.104806 -0.219883 -0.342195 -0.650833  0.858232  1.101332  
4 -0.229632 -0.432313  0.099920 -0.318309  1.409422 -0.395228  

[5 rows x 52 columns]

# 模型搭建
先直接对Class预测

In [32]:
# 数据集分割
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score, accuracy_score, log_loss
x_train, x_test, y_train, y_test = train_test_split(train, ytrain, test_size=0.2, random_state=42)
scale_pos_weight = 4.712962962962963
print(x_train.shape)
print(y_train.shape)

(493, 52)
(493, 5)


In [33]:
def ScoreMetric(ytrue, ypred):
    nc = np.bincount(ytrue);
    return log_loss(ytrue, ypred, sample_weight = 1 / nc[ytrue], eps=1e-15);

def lgb_metric(y_true, y_pred):
    return 'balanced_log_loss', ScoreMetric(y_true, y_pred), False

In [34]:
# # xgb
import xgboost as xgb
xgb_c = xgb.XGBClassifier(n_estimators=300, learning_rate=0.01, max_depth=4, random_state=42)
xgb_c.fit(x_train, y_train['Class'])
xgb_pred = xgb_c.predict(x_test)
y_pred = xgb_c.predict_proba(x_test)
xgb_accuracy = accuracy_score(xgb_pred, y_test['Class'])
# xgb_pred
xgb_accuracy

0.967741935483871

In [35]:
p0 = y_pred[:,1]
ScoreMetric(y_test['Class'], p0)
# p0

0.3443425278858093

# 正式的KFold交叉验证 + Optuna模型调参

In [36]:
import optuna, tune

from optuna.samplers import TPESampler
from optuna.visualization import plot_contour
from optuna.visualization import plot_edf
from optuna.visualization import plot_intermediate_values
from optuna.visualization import plot_optimization_history
from optuna.visualization import plot_parallel_coordinate
from optuna.visualization import plot_param_importances
from optuna.visualization import plot_slice

from sklearn.model_selection import GridSearchCV, cross_val_score, cross_validate
from sklearn.metrics import confusion_matrix, roc_curve, roc_auc_score, log_loss
from sklearn.metrics import make_scorer, accuracy_score, log_loss
from sklearn.model_selection import ShuffleSplit, StratifiedShuffleSplit

# fold
from sklearn.model_selection import KFold, StratifiedKFold


In [37]:
def objective(trial):
    # dtrain = xgb.DMatrix(train, label=ytrain['Class'])
    '''
    Value: 0.2754406155516159
    Params: 
        booster: gbtree
        lambda: 0.24302130237467398
        alpha: 2.2108795894332237e-08
        subsample: 0.7887998109844777
        colsample_bytree: 0.4493748858737539
        max_depth: 2
        eta: 0.29846902305256606
        gamma: 1.4539411503015061e-05
        grow_policy: lossguide
    '''
    param = {
        "verbosity": 0,
        "objective": "binary:logistic",
        "booster": trial.suggest_categorical("booster", ["gbtree", "gblinear", "dart"]),
        "lambda": trial.suggest_float("lambda", 1e-8, 1.0, log=True),
        "alpha": trial.suggest_float("alpha", 1e-8, 1.0, log=True),
        "subsample": trial.suggest_float("subsample", 0.2, 0.8),
        "colsample_bytree": trial.suggest_float("colsample_bytree", 0.2, 0.8),
        'scale_pos_weight' : scale_pos_weight,
        'random_state' : 42,
        "learning_rate": trial.suggest_float("learning_rate", 0.01, 0.3)
    }

    if param["booster"] == "gbtree" or param["booster"] == "dart":
        param["max_depth"] = trial.suggest_int("max_depth", 1, 16)
        param["eta"] = trial.suggest_float("eta", 1e-3, 1.0, log=True)
        param["gamma"] = trial.suggest_float("gamma", 1e-8, 1.0, log=True)
        param["grow_policy"] = trial.suggest_categorical("grow_policy", ["depthwise", "lossguide"])
    if param["booster"] == "dart":
        param["sample_type"] = trial.suggest_categorical("sample_type", ["uniform", "weighted"])
        param["normalize_type"] = trial.suggest_categorical("normalize_type", ["tree", "forest"])
        param["rate_drop"] = trial.suggest_float("rate_drop", 1e-8, 1.0, log=True)
        param["skip_drop"] = trial.suggest_float("skip_drop", 1e-8, 1.0, log=True)

    skf = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)
    log_loss_scores = []
    # 划分验证集
    for train_index, val_index in skf.split(train, ytrain['Class']):
        x_train_, x_val = train.iloc[train_index], train.iloc[val_index]
        y_train_, y_val = ytrain['Class'].iloc[train_index], ytrain['Class'].iloc[val_index]
        classifier = xgb.XGBClassifier(**param, verbose=0)
        classifier.fit(x_train_, y_train_, eval_set=[(x_val, y_val)], early_stopping_rounds=180, verbose=0, scale_pos_weight=scale_pos_weight)
        y_pred_proba = classifier.predict_proba(x_val)[:, 1]
        log_loss_scores.append(ScoreMetric(y_val, y_pred_proba))
    return np.mean(log_loss_scores)

if __name__ == "__main__":
    sampler = TPESampler(seed=42)
    pruner = optuna.pruners.MedianPruner(n_warmup_steps=10)
    study = optuna.create_study(pruner=pruner, direction="minimize", sampler=sampler)
    study.optimize(objective, n_trials=500)

    print("Number of finished trials: {}".format(len(study.trials)))
    print("Best trial:")
    trial = study.best_trial
    print("  Value: {}".format(trial.value))
    print("  Params: ")
    for key, value in trial.params.items():
        print("    {}: {}".format(key, value))

[I 2023-08-02 22:44:57,678] A new study created in memory with name: no-name-36a7d3e6-b74d-4c07-87b1-293592a2417e
[W 2023-08-02 22:44:58,373] Trial 0 failed with parameters: {'booster': 'gblinear', 'lambda': 0.0006155564318973012, 'alpha': 1.77071686435378e-07, 'subsample': 0.2935967122017216, 'colsample_bytree': 0.23485016730091968, 'learning_rate': 0.2611910822747312} because of the following error: The value nan is not acceptable..
[W 2023-08-02 22:44:58,374] Trial 0 failed with value nan.
[I 2023-08-02 22:44:59,081] Trial 1 finished with value: 0.519294173810417 and parameters: {'booster': 'gblinear', 'lambda': 0.574485163632042, 'alpha': 0.04566054873446119, 'subsample': 0.3274034664069657, 'colsample_bytree': 0.3090949803242604, 'learning_rate': 0.06318730785749581}. Best is trial 1 with value: 0.519294173810417.
[W 2023-08-02 22:44:59,785] Trial 2 failed with parameters: {'booster': 'gblinear', 'lambda': 2.1371407316372935e-06, 'alpha': 0.000784915956255507, 'subsample': 0.28369

Number of finished trials: 500
Best trial:
  Value: 0.2535332401012661
  Params: 
    booster: dart
    lambda: 0.0067981943191443815
    alpha: 1.1158232780616973e-07
    subsample: 0.39680328510099894
    colsample_bytree: 0.5482579412080295
    learning_rate: 0.13779959534970157
    max_depth: 3
    eta: 0.0010017919482385215
    gamma: 0.030524056409888256
    grow_policy: depthwise
    sample_type: weighted
    normalize_type: tree
    rate_drop: 4.4093900571337765e-06
    skip_drop: 1.105656318625032e-05


In [38]:
model = xgb.XGBClassifier(**trial.params)
model.fit(x_train, y_train['Class'])
y_pred = model.predict(x_test)
y_pred_proba = model.predict_proba(x_test)[:, 1]
ScoreMetric(y_test['Class'], y_pred_proba)
accuracy_score(y_pred, y_test['Class'])

0.9193548387096774

In [39]:
plot_optimization_history(study)

In [40]:
plot_slice(study)

In [41]:
plot_param_importances(study)

In [42]:
trial.params

{'booster': 'dart',
 'lambda': 0.0067981943191443815,
 'alpha': 1.1158232780616973e-07,
 'subsample': 0.39680328510099894,
 'colsample_bytree': 0.5482579412080295,
 'learning_rate': 0.13779959534970157,
 'max_depth': 3,
 'eta': 0.0010017919482385215,
 'gamma': 0.030524056409888256,
 'grow_policy': 'depthwise',
 'sample_type': 'weighted',
 'normalize_type': 'tree',
 'rate_drop': 4.4093900571337765e-06,
 'skip_drop': 1.105656318625032e-05}

1.
{'booster': 'gbtree',
 'lambda': 0.24302130237467398,
 'alpha': 2.2108795894332237e-08,
 'subsample': 0.7887998109844777,
 'colsample_bytree': 0.4493748858737539,
 'max_depth': 2,
 'eta': 0.29846902305256606,
 'gamma': 1.4539411503015061e-05,
 'grow_policy': 'lossguide'}

 2.
 {'booster': 'gbtree',
 'lambda': 0.8068353722333764,
 'alpha': 1.894027813634802e-08,
 'subsample': 0.7785041239304065,
 'colsample_bytree': 0.7601163387370006,
 'learning_rate': 0.02,
 'max_depth': 4,
 'eta': 0.645702469472196,
 'gamma': 3.1878972195087093e-07,
 'grow_policy': 'lossguide'}

 3.
 {'booster': 'dart',
 'lambda': 0.0067981943191443815,
 'alpha': 1.1158232780616973e-07,
 'subsample': 0.39680328510099894,
 'colsample_bytree': 0.5482579412080295,
 'learning_rate': 0.13779959534970157,
 'max_depth': 3,
 'eta': 0.0010017919482385215,
 'gamma': 0.030524056409888256,
 'grow_policy': 'depthwise',
 'sample_type': 'weighted',
 'normalize_type': 'tree',
 'rate_drop': 4.4093900571337765e-06,
 'skip_drop': 1.105656318625032e-05}